core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -727,12 +732,21 @@ private Type handleResolvedField(Expression node, ResolvedField resolvedField, S
 
             FieldId fieldId = FieldId.from(resolvedField);
             Field field = resolvedField.getField();
+            Type fieldType = field.getType();
+            Optional<Node> sourceNode = fieldId.getRelationId().getSourceNode();
+            if (sourceNode.isPresent() && sourceNode.get() instanceof Relation relation) {
+                List<Type> coercions = getRelationCoercion.apply(relation);
```

Comment:
This doesn’t look right. From the point of view of the expression analyzer, the field should already have the right type based on those coercions - and recorded in the scope.

Reply:
The issue is here https://github.com/trinodb/trino/blob/92c74d85dc01a073e0ae985e3dc13ddb0e0b695c/core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java#L2572
```
            if (storageTable.isPresent()) {
                List<Field> storageTableFields = analyzeStorageTable(table, viewFields, storageTable.get());
                Scope accessControlScope = Scope.builder()
                        .withRelationType(RelationId.anonymous(), new RelationType(viewFields))
                        .build();
                analyzeFiltersAndMasks(table, name, new RelationType(viewFields), accessControlScope);
                analysis.registerTable(table, storageTable, name, session.getIdentity().getUser(), accessControlScope, Optional.of(originalSql));
                analysis.addRelationCoercion(table, viewFields.stream().map(Field::getType).toArray(Type[]::new));
                // use storage table output fields as they contain ColumnHandles
                return createAndAssignScope(table, scope, storageTableFields);
            }
```
`storageTableFields` don't have coerced types, but the underlying table is already analyzed with them, so we cannot just swap them for `viewFields`.

Another idea is to swap type when creating `Field` in `analyzeStorageTable`, but it feels more hacky than this PR as `analyzeTableOutputFields` would need to be changed too.

Reply:
To add to this, `getRelationCoercion` is never called between that above-linked line and where the exception is thrown in `ExpressionAnalyzer`. It is only used in planning, not analysis, which is essentially the root cause of this bug.

@martint do you have a recommendation for a better way to solve this? I've tried coercing the fields of the `Scope` passed to `analyzeExpression` from `analyzeWhere`:
```java
Scope expressionScope = scope;
Optional<Node> sourceNode = scope.getRelationId().getSourceNode();
if (sourceNode.isPresent() && sourceNode.get() instanceof Relation relation) {
    List<Type> coercion = analysis.getRelationCoercion(relation);
    if (coercion != null) {
        List<Field> coercedFields = coerceFields(ImmutableList.copyOf(scope.getRelationType().getAllFields()), coercion);
        expressionScope = scopeBuilder(Optional.of(scope))
                .withRelationType(RelationId.of(node), new RelationType(coercedFields))
                .build();
    }
}
ExpressionAnalysis expressionAnalysis = analyzeExpression(predicate, expressionScope);
```
but this ends up yielding the following planner error
```
java.lang.IllegalStateException: Unexpected identifier in logical plan: tswtz_3
	at io.trino.sql.planner.sanity.NoIdentifierLeftChecker.validate(NoIdentifierLeftChecker.java:42)
	at io.trino.sql.planner.sanity.PlanSanityChecker.lambda$validate$0(PlanSanityChecker.java:129)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:422)
	at io.trino.sql.planner.sanity.PlanSanityChecker.validate(PlanSanityChecker.java:129)
	at io.trino.sql.planner.sanity.PlanSanityChecker.validateIntermediatePlan(PlanSanityChecker.java:94)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:280)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:257)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:252)
	at io.trino.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:540)
	at io.trino.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:518)
	at io.trino.execution.SqlQueryExecution.start(SqlQueryExecution.java:456)
	at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:273)
	at io.trino.dispatcher.LocalDispatchQuery.startExecution(LocalDispatchQuery.java:148)
	at io.trino.dispatcher.LocalDispatchQuery.lambda$waitForMinimumWorkers$2(LocalDispatchQuery.java:132)
	at io.airlift.concurrent.MoreFutures.lambda$addSuccessCallback$12(MoreFutures.java:568)
	at io.airlift.concurrent.MoreFutures$3.onSuccess(MoreFutures.java:543)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1133)
	at io.trino.$gen.Trino_testversion____20240227_211140_75.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
	Suppressed: java.lang.Exception: Current plan:
                Output[columnNames = [t_3, t_9, ts_3, ts_9, tswtz_3, tswtz_9]]
                │   Layout: [t:time(6), t_0:time(6), ts:timestamp(6), ts_1:timestamp(6), tswtz_3:timestamp(6) with time zone, tswtz_4:timestamp(6) with time zone]
                │   t_3 := t
                │   t_9 := t_0
                │   ts_3 := ts
                │   ts_9 := ts_1
                │   tswtz_9 := tswtz_4
                └─ Project[]
                   │   Layout: [t:time(6), t_0:time(6), ts:timestamp(6), ts_1:timestamp(6), tswtz_3:timestamp(6) with time zone, tswtz_4:timestamp(6) with time zone]
                   └─ Project[]
                      │   Layout: [t:time(6), t_0:time(6), ts:timestamp(6), ts_1:timestamp(6), tswtz_3:timestamp(6) with time zone, tswtz_4:timestamp(6) with time zone]
                      └─ Project[]
                         │   Layout: [t:time(6), t_0:time(6), ts:timestamp(6), ts_1:timestamp(6), tswtz_3:timestamp(6) with time zone, tswtz_4:timestamp(6) with time zone]
                         └─ FilterProject[filterPredicate = (tswtz_3 < CAST("$current_timestamp"(CAST(null AS timestamp(3) with time zone)) AS timestamp(6) with time zone))]
                            │   Layout: [t:time(6), t_0:time(6), ts:timestamp(6), ts_1:timestamp(6), tswtz_3:timestamp(6) with time zone, tswtz_4:timestamp(6) with time zone]
                            └─ ScanProject[table = iceberg:default.timestamp_mv_test$materialized_view_storage$data@2177961468857097358]
                                   Layout: [t:time(6), t_0:time(6), ts:timestamp(6), ts_1:timestamp(6), tswtz_3:timestamp(6) with time zone, tswtz_4:timestamp(6) with time zone]
                                   tswtz_3 := CAST("tswtz" AS timestamp(6) with time zone)
                                   tswtz_4 := CAST("tswtz_2" AS timestamp(6) with time zone)
                                   t := 1:t_3:time(6)
                                   t_0 := 2:t_9:time(6)
                                   ts := 3:ts_3:timestamp(6)
                                   ts_1 := 4:ts_9:timestamp(6)
                                   tswtz := 5:tswtz_3:varchar
                                   tswtz_2 := 6:tswtz_9:varchar

		at io.trino.sql.planner.sanity.PlanSanityChecker.validate(PlanSanityChecker.java:143)
		... 17 more
```

Reply:
The proper fix is to:

During analysis:
1) analyze the storage table
2) record a coercion for the fields if necessary
3) create a scope based the materialized view fields
4) record the linkage between the MV and the storage table

During planning, when planning the scan of the Table reference
1) If there is a recorded link between MV -> storage table, plan the storage table
   a) Insert a projection to perform the coercions
2) Create a subplan with the MV fields & scope as baseline, to correspond with the fields of the coerced storage table subplan.

Most of that logic already exists, so it shouldn't be a huge delta from what's there. It may need some additional information in the Analysis object to record the linkage between MV -> storage table.

Reply:
@martint 

 > record the linkage between the MV and the storage table

Why link from MV to storage table is cleaner than link from storage table to right type via coercion?

Reply:
The scope needs to expose the fields from the view so that they have the correct type. The planner needs to produce a subplan for that Table AST node that matches the schema of the MV.

Whether it’s encoded as coercion info  the planner uses to produce that subplan or an explicit recording of the indirection is an detail to be figured out, but that’s the essence of how the planning process should work.

In a sense, when the MV has a storage table, the MV should be planned as if it were a view over “select … from storage_table” with any necessary casts inserted in the select clause to adjust the types.

Reply:
Here's a potential fix: https://github.com/martint/trino/commit/bbb81ebf9c67dd12d267f707391bd81dd1c5ee11

Reply:
@martint thank you very much for your help - i took on your suggestions, including the removal of addRelationCoercion